### PR TITLE
Refactor Data Pipeline

### PR DESCRIPTION
**Pull request by nikmo33**
From a conversation with @ethanweber decided to take https://github.com/nerfstudio-project/nerfstudio/pull/886 a step further and cleanup input metadata overall and make this pipeline more generic. Adds input metadata that may be useful for adding new input modalities in the future.
Left mask as a separate variable from metadata as it is one that is probably something we want to support in the base dataset? 
Added an example usage with semantic nerfw and associated data code

Move datamanagers into their own folder and split into base/semantic datamanagers
Move datasets into their own folder and split into base/semantic datasets

Have tested with vanilla nerfacto with poster dataset and semantic-nerfw with friends-data. Also checked viewer and everything seems to be working there + ns-dev-test

### CODE REVIEW

File: nerfstudio/data/pixel_samplers.py [hunk 1]
```diff
@@ -48,30 +48,19 @@ def collate_image_dataset_batch(batch: Dict, num_rays_per_batch: int, keep_full_
...
-    }
-    if mask is not None:
-        collated_batch["mask"] = mask
-    if semantics_stuff is not None:
-        collated_batch["semantics_stuff"] = semantics_stuff
-    if semantics_thing is not None:
-        collated_batch["semantics_thing"] = semantics_thing
+    collated_batch.update(
+        {
+            "local_indices": local_indices,  # local to the batch returned
```
Comment (ethanweber) [inline nerfstudio/data/pixel_samplers.py:60]: I guess we should really just delete this. It's not used anywhere in the code anymore. We used it previously, but it's unnecessary.

File: nerfstudio/pipelines/base_pipeline.py [hunk 2]
```diff
@@ -220,8 +220,7 @@ def __init__(
         self._model = config.model.setup(
             scene_box=self.datamanager.train_dataset.dataparser_outputs.scene_box,
             num_train_data=len(self.datamanager.train_dataset),
-            semantics=self.datamanager.train_dataset.dataparser_outputs.semantics,
-            device=device,
+            metadata=self.datamanager.train_dataset.dataparser_outputs.metadata,
```
Comment (ethanweber) [inline nerfstudio/pipelines/base_pipeline.py:223]: Looks good for this PR. I do wish, however, that this was cleaner. I guess the better thing to do is probably pass the DataParserOutputs into all the Model classes? Not too sure at the moment.

**Review by ethanweber — APPROVED** [review]
Looks great. Thanks for making these changes. Left a few comments but quite minor overall.

File: nerfstudio/data/datasets/base_dataset.py [hunk 3]
```diff
@@ -80,13 +82,23 @@ def get_data(self, image_idx) -> Dict:
...
-        for _, data_func_dict in self.dataparser_outputs.additional_inputs.items():
-            assert "func" in data_func_dict, "Missing function to process data: specify `func` in `additional_inputs`"
-            func = data_func_dict["func"]
-            assert "kwargs" in data_func_dict, "No data to process: specify `kwargs` in `additional_inputs`"
-            data.update(func(image_idx, **data_func_dict["kwargs"]))
+        if self.has_masks:
+            mask_filepath = self.dataparser_outputs.mask_filenames[image_idx]
+            data["mask"] = get_image_mask_tensor_from_path(filepath=mask_filepath)
+        metadata = self.get_metadata(data)
+        data.update(metadata)
```
[the remark this answers is not in the document]
Comment (nikmo33) [reply nerfstudio/data/datasets/base_dataset.py:89]: yup Ive tested this locally with some custom data

**Comment by nikmo33** [comment]
> Looks great. Thanks for making these changes. Left a few comments but quite minor overall.

Thanks! Comments should be addressed

**Comment by ethanweber** [comment]
LGTM!

**Comment by nikmo33** [comment]
@ethanweber I dont think i can merge this myself 😛
